### DISCUSSION TITLE
MeteorJs with SOLID Cloud infrastructure

### DISCUSSION THREAD

**Discussion by KheireddineAzzez** [meteor/meteor#12191] [Feature Requests]
I was wondering lately why meteor does not provide a completely **independent cloud infrastructure** or does not ensure a good integration between **existing ones.**

For example : ### AWS(S3...), MINIO or even AZURE.

As this will ensure cohesion between containers when we use the load balancing process, moreover, it will be amazing when it provides a complete visualization of what is going on inside the load balancer; because the power of Meteor is based on the DDP protocol and compatibility with Front-end multidisciplinary frameworks. So it will be very difficult to simple developer  to insure something like that from scratch.  Besides that, there are many packages in AtmosphereJS which are antiquated and useless.

It will be magnificent  to see that kind of feature in the near future.

**Comment by xet7** [comment]
@KheireddineAzzez 

With S3 and Minio, do you mean https://github.com/veliovgroup/Meteor-Files ?

There are also Meteor Community packages:
- https://github.com/Meteor-Community-Packages
- https://dev.to/storytellercz/what-is-the-meteor-community-packages-community-nc0

Many Meteor apps are running at AWS, Azure, etc, with running fron source bundle, Docker, Snap, etc.

What visualization tool do you mean? Some already existing at Meteor Cloud, or elsewhere?

What kind of working package for Meteor you are trying to find ?

**Reply by KheireddineAzzez** [reply]
> @KheireddineAzzez
> 
> With S3 and Minio, do you mean https://github.com/veliovgroup/Meteor-Files ?
> 
> There are also Meteor Community packages:
> 
> * https://github.com/Meteor-Community-Packages
> * https://dev.to/storytellercz/what-is-the-meteor-community-packages-community-nc0
> 
> Many Meteor apps are running at AWS, Azure, etc, with running fron source bundle, Docker, Snap, etc.
> 
> What visualization tool do you mean? Some already existing at Meteor Cloud, or elsewhere?
> 
> What kind of working package for Meteor you are trying to find ?

Thank you for the response, my idea was about an enhancement that can allow the developer to  freely choose the type of  database.
For example if he want to use REDIS as a cache database or choose the LOADBALANCER on which he want to execute his project like docker .

**Reply by xet7** [reply]
@KheireddineAzzez 

With redis, do you mean https://github.com/cult-of-coders/redis-oplog ?

With other database, do you mean PostgreSQL ? https://github.com/meteor/meteor/discussions/12150

**Reply by xet7** [reply]
What kind of loadbalancer you mean? At AWS it usually is ELB. Also Traefik, Nginx, Caddy, Apache, HAProxy could be used.

**Comment by KheireddineAzzez** [comment]
I am talking  about haproxy Nginx...

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Reply by xet7** [reply]
@KheireddineAzzez 

## Deploying Meteor

https://guide.meteor.com/deployment.html

## Webservers

- Nginx https://github.com/wekan/wekan/wiki/Nginx-Webserver-Config
- Caddy https://github.com/wekan/wekan/wiki/Caddy-Webserver-Config
- Apache https://github.com/wekan/wekan/wiki/Apache
- Traefik https://github.com/wekan/wekan/wiki/Traefik-and-self-signed-SSL-certs
- OpenLiteSpeed https://github.com/wekan/wekan/issues/3334#issuecomment-723651328

## Proxying

For proxying from TLS webserver (like some of above) to many instances of running Meteor Node.js HTTP ports running on different servers, add those proxy IP and port addresses to webserver config. It could maybe need sticky sessions, info could be found from some search maybe.

## Docker

Some settings for for some Meteor app (in this case WeKan https://wekan.github.io ) 

https://github.com/wekan/wekan/issues/4704#issuecomment-1260018694

**Reply by KheireddineAzzez** [reply]
[Quick diagram](https://github.com/meteor/meteor/files/9662707/MeteorJS.pdf)


**Description** :
 we all know that meteor is running on NodeJS.

- The **adapter** is used to keep clean architecture ( Hexagonal) and let the developer focus only on the logic part.
     So in this case, the project manager will choose the perfect database to work with.
     Besides that, everything will be SOLID and easy to maintain.

- All projects(DatatBases, MeteorJS, Cache DataBase) will be hosted on GitLab, for that reason, it will be an  interface created for the developer, so the developer will need only to choose the type of database that he wants to work with, and by one click everything will be ready to be uploaded. Additionally, the developer will be able to see the progress of his work, and he will be able to maintain scalability if there is a lack of resources.

**Reply by xet7** [reply]
There are many parts in Meteor that work only with MongoDB: At server, MongoDB 5.x (in future 6.x), and at client minimongo (database implemented with Javascript). MongoDB 6 will bring possibility to search encrypted database etc. That [OxideDB/FerretDB](https://github.com/meteor/meteor/discussions/12150) only proxies that to PostgreSQL (and maybe in future other SQL databases).

Realtime UI in Meteor works by reading MongoDB change stream newest changes, and updating that to all logged in users, with Javascript that runs at client, server, or both. Where it runs, is defined in code like `Meteor.isServer()` or `Meteor.isClient()`, or having code outside of that. Meteor can use PubSub to publish selected fields of MongoDB data to clientside. By starting and stopping those subscriptions, it's possible to change what amount of data is updated in realtime clientside. If for some reason there is too much Javascript clientside, it could be moved to run serverside.

Meteor can use packages from both atmospherejs.com and npmjs.com . So it is possible to:

a) Add package like TypeORM (or similar ORM / DAL) that would use TypeORM syntax to connect to any database it supports

b) Add any other npm database driver there is at npmjs.com or elsewhere

c) Use Javascript to connect to any other service made with any programming language and database, to use that data 

Scalability is usually done by separating to different services at different servers. For example, if someone at web UI click something , it could use Meteor Method, that at serverside checks required permissions, and calls with REST API some external service that does long running batch jobs. While Meteor has many parts for auth, email, and other functionality, but if something else is needed, that could be called with REST API etc.

For frontend Javascript web framework, Meteor can use various frameworks. Command `meteor create` shows these options:

```
Options:
  --package     Create a new meteor package instead of an app.
  --example     Example template to use.
  --list        Show list of available examples.
  --bare        Create an empty app.
  --minimal     Create an app with as few Meteor packages as possible.
  --full        Create a fully scaffolded app.
  --react       Create a basic react-based app, same as default.
  --vue         Create a basic vue-based app.
  --apollo      Create a basic apollo-based app.
  --svelte      Create a basic svelte-based app.
  --typescript  Create a basic Typescript React-based app.
  --blaze       Create a basic blaze-based app.
  --tailwind    Create a basic react-based app, with tailwind configured.
```

Or alternatively, use only serverside rendering https://github.com/wekan/we .

It kind of depends, what parts of Meteor is needed, and to what purpose. Then looking at bottlenecks, and separating those to external services.

I just presume, that maintaining is not so easy, be it SOLID, Meteor, or anything else. Maintaining is a lot of work.

**Reply by xet7** [reply]
But, those added npm packages like TypeORM etc would not be so integrated, when compared to current Meteor with MongoDB. I have not tried it yet.
